### PR TITLE
RavenDB-18018 - [Sharding] Get studio footer stats (shard wide)

### PR DESCRIPTION
**Pull request by lastav5**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18018

### CODE REVIEW

File: src/Raven.Server/Documents/ShardedHandlers/Processors/ShardedStatsHandlerProcessorForGetStudioFooterStats.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        public ShardedStatsHandlerProcessorForGetStudioFooterStats([NotNull] ShardedRequestHandler requestHandler) : base(requestHandler, requestHandler.ContextPool)
+        {
+        }
+        
+        protected override async ValueTask<FooterStatistics> GetFooterStatistics()
+        {
+            var op = new ShardedStudioStatsHandler.ShardedGetStudioFooterStatsOperation();
+            var stats = await RequestHandler.ShardExecutor.ExecuteParallelForAllAsync(op);
+            stats.CountOfIndexes = RequestHandler.ShardedContext.DatabaseRecord.Indexes.Count;
+            //missing CountOfStaleIndexes and CountOfIndexingErrors fields
```
Comment (karmeli87) [inline src/Raven.Server/Documents/ShardedHandlers/Processors/ShardedStatsHandlerProcessorForGetStudioFooterStats.cs]: please note this in the ticket itself

**Review by karmeli87 — APPROVED** [review]

File: src/Raven.Server/Documents/ShardedHandlers/ShardedStudioStatsHandler.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        [RavenShardedAction("/databases/*/studio/footer/stats", "GET")]
+        public async Task FooterStats()
+        {
+            using (var processor = new ShardedStatsHandlerProcessorForGetStudioFooterStats(this))
+            {
+                await processor.ExecuteAsync();
+            }
+        }
+
+        public readonly struct ShardedGetStudioFooterStatsOperation : IShardedOperation<FooterStatistics>
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedStudioStatsHandler.cs]: put that inside `ShardedStatsHandlerProcessorForGetStudioFooterStats`

File: src/Raven.Server/Web/Studio/StudioStatsHandler.cs [hunk 3]
```diff
@@ -13,6 +14,11 @@ public class StudioStatsHandler : DatabaseRequestHandler
         [RavenAction("/databases/*/studio/footer/stats", "GET", AuthorizationStatus.ValidUser, EndpointType.Read)]
         public async Task FooterStats()
         {
+            using (var processor = new StatsHandlerProcessorForGetStudioFooterStats(this))
+            {
+                await processor.ExecuteAsync();
+            }
+
             using (var context = QueryOperationContext.Allocate(Database, needsServerContext: true))
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/StudioStatsHandler.cs]: we need to remove this, right?

File: src/Raven.Server/Documents/Handlers/Processors/AbstractStatsHandlerProcessorForGetStudioFooterStats.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+{
+    internal abstract class AbstractStatsHandlerProcessorForGetStudioFooterStats<TRequestHandler, TOperationContext> : AbstractHandlerProcessor<TRequestHandler, TOperationContext>
+        where TRequestHandler : RequestHandler
+        where TOperationContext : JsonOperationContext
+    {
+        public AbstractStatsHandlerProcessorForGetStudioFooterStats([NotNull] TRequestHandler requestHandler, [NotNull] JsonContextPoolBase<TOperationContext> contextPool) : base(requestHandler, contextPool)
+        {
+        }
+
+        protected abstract ValueTask<FooterStatistics> GetFooterStatistics();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/AbstractStatsHandlerProcessorForGetStudioFooterStats.cs]: Async methods need to have `Async` at the end

File: test/SlowTests/SlowTests/StudioTests.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                Assert.Equal(null, stats.CountOfIndexingErrors);
+                Assert.Equal(null, stats.CountOfStaleIndexes);
+            }
+        }
+
+        private class User
+        {
+            public string Name;
+
+            public bool Banned;
```
Comment (ppekrol) [inline test/SlowTests/SlowTests/StudioTests.cs]: do you need that field?

File: test/SlowTests/SlowTests/StudioTests.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                }
+                await store.ExecuteIndexAsync(new AllowedUsers());
+                WaitForIndexing(store, sharded: true);
+
+                var stats = store.Maintenance.Send(new GetStudioFooterStatisticsOperation());
+
+                Assert.NotNull(stats);
+                Assert.Equal(1, stats.CountOfIndexes);
+                Assert.Equal(2, stats.CountOfDocuments);
+                Assert.Equal(null, stats.CountOfIndexingErrors);
```
Comment (ppekrol) [inline test/SlowTests/SlowTests/StudioTests.cs]: when you will rebase Options will have 2 additional properties. DatabaseMode and SearchEngineMode. Use them to do a proper assert and merge the test.

File: src/Raven.Server/Documents/ShardedHandlers/ShardedStudioStatsHandler.cs (hunk 2 above)
Comment (lastav5) [reply src/Raven.Server/Documents/ShardedHandlers/ShardedStudioStatsHandler.cs]: done

File: src/Raven.Server/Web/Studio/StudioStatsHandler.cs (hunk 3 above)
Comment (lastav5) [reply src/Raven.Server/Web/Studio/StudioStatsHandler.cs]: yes. removed

File: src/Raven.Server/Documents/Handlers/Processors/AbstractStatsHandlerProcessorForGetStudioFooterStats.cs (hunk 4 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/AbstractStatsHandlerProcessorForGetStudioFooterStats.cs]: changed

File: test/SlowTests/SlowTests/StudioTests.cs (hunk 5 above)
Comment (lastav5) [reply test/SlowTests/SlowTests/StudioTests.cs]: not really. removed

File: test/SlowTests/SlowTests/StudioTests.cs (hunk 6 above)
Comment (lastav5) [reply test/SlowTests/SlowTests/StudioTests.cs]: done

File: src/Raven.Server/Documents/ShardedHandlers/Processors/ShardedStatsHandlerProcessorForGetStudioFooterStats.cs (hunk 1 above)
Comment (lastav5) [reply src/Raven.Server/Documents/ShardedHandlers/Processors/ShardedStatsHandlerProcessorForGetStudioFooterStats.cs]: noted

**Pull request merged**